File: kotlin/src/main/kotlin/advertisements/advertisement/infrastructure/stream/producer/AdvertisementEventsProducer.kt
package advertisements.advertisement.infrastructure.stream.producer

import advertisements.advertisement.domain.events.AdvertisementWasApproved
import advertisements.advertisement.infrastructure.stream.producer.events.AdvertisementApprovedEvent
import common.domain.DomainEvent
import common.domain.EventPublisher
import common.infrastructure.stream.MessageBroker
import common.infrastructure.stream.producer.SerializableEvent
import framework.ThreadContext
import java.util.UUID

class AdvertisementEventsProducer (
    private val messageBroker: MessageBroker
): EventPublisher {
    companion object {
        const val PUB_ADVERTISEMENT = "pub.advertisement"
    }

    override fun publish(events: List<DomainEvent>) {
        for (event in events) {
            publishToMessageBroker(event)
        }
    }

    private fun publishToMessageBroker(event: DomainEvent) {
        try {
            when (event) {
                is AdvertisementWasApproved -> publishAdvertisementApproved(event)
                else -> {} // Handle other event types or do nothing
            }
        } catch (e: Exception) {
            throw RuntimeException("Error publishing event to message broker", e)
        }
    }

    private fun publishAdvertisementApproved(event: AdvertisementWasApproved) {
        val correlationId = ThreadContext.get("correlationId")?.toString() ?: UUID.randomUUID().toString()
        val causationId = ThreadContext.get("causationId")?.toString()

        val advertisementApprovedEvent = AdvertisementApprovedEvent.create(
            event,
            correlationId,
            causationId
        )
        sendEventToMessageBroker(advertisementApprovedEvent)
    }

    private fun sendEventToMessageBroker(event: SerializableEvent) {
        messageBroker.publish(event, PUB_ADVERTISEMENT)
    }
}

File: typescript/src/advertisements/advertisement/infrastructure/stream/producer/AdvertisementEventsProducer.ts
import {AdvertisementWasApproved} from "../../../domain/events/AdvertisementWasApproved";
import {AdvertisementApprovedEvent} from "./events/AdvertisementApprovedEvent";
import {DomainEvent} from "../../../../../common/domain/DomainEvent";
import {EventPublisher} from "../../../../../common/domain/EventPublisher";
import {SerializableEvent} from "../../../../../common/infrastructure/stream/SerializableEvent";
import {MessageBroker} from "../../../../../common/infrastructure/stream/MessageBroker";

export class AdvertisementEventsProducer implements EventPublisher {
    private static readonly PUB_ADVERTISEMENT: string = 'pub.advertisement';

    constructor(private messageBroker: MessageBroker) {}

    public publish(...events: DomainEvent[]): void {
        for (const event of events) {
            this.publishToMessageBroker(event);
        }
    }

    private publishToMessageBroker(event: DomainEvent): void {
        try {
            if (event instanceof AdvertisementWasApproved) {
                this.publishAdvertisementApproved(event);
            }
        } catch (e) {
            throw new Error('Error publishing event to message broker');
        }
    }

    private publishAdvertisementApproved(event: AdvertisementWasApproved): void {
        const approvedEvent = AdvertisementApprovedEvent.create(event);
        this.sendEventToMessageBroker(approvedEvent);
    }

    private sendEventToMessageBroker(event: SerializableEvent): void {
        this.messageBroker.publish(event, AdvertisementEventsProducer.PUB_ADVERTISEMENT);
    }
}